### ISSUE TITLE
[4.0.0-beta.3] webSecurity in webPreferences ignored (CORB)

### ISSUE THREAD

**Issue by IAmPicard** [electron/electron#15132]
* Output of `node_modules/.bin/electron --version`: **v4.0.0-beta.3**
* Operating System (Platform and Version): **Windows 10**
* Output of `node_modules/.bin/electron --version` on last known working Electron version (if applicable): **v3.0.4**

**Expected Behavior**
With electron 3, I could pass `webPreferences: {webwebSecurity: false}` to the BrowserWindow constructor (the equivalent of the `--disable-web-security` Chrome command line argument) which would disable CORB (effectively allowing cross-origin requests).

**Actual behavior**
CORB is still doing its thing (ignoring the `{webwebSecurity: false}` option; I see entries like these in the console:
`Cross-Origin Read Blocking (CORB) blocked cross-origin response <URL> with MIME type application/json. See <URL> for more details.`

**To Reproduce**
Make a request (using fetch or xhr) to a file on a different domain.
Example repo:
```
git clone  https://github.com/IAmPicard/electron-cors-repro
npm install
npm run start
```

**Comment by blackknifes** [comment]
@MarshallOfSound 
The bug also is exists on version 3.x.

**Comment by MarshallOfSound** [comment]
@blackknifes That directly contradicts the statement made in the original issue

> With electron 3, I could pass webPreferences: {webwebSecurity: false} to the BrowserWindow constructor (the equivalent of the --disable-web-security Chrome command line argument) which would disable CORB (effectively allowing cross-origin requests).

@IAmPicard Can you confirm v3 works as expected?

**Comment by IAmPicard** [comment]
Yes, everything worked as expected ever since I started using it (back in the 1.8 days) up-to the latest available 3.0 (3.0.4).

**Comment by nornagon** [comment]
Is it possible you could put together a small repro case (perhaps with https://github.com/electron/fiddle)? I'm not sure exactly how to trigger CORB and I'd like to make sure I'm testing exactly what you're seeing.

**Comment by blackknifes** [comment]
@IAmPicard 
When I started with vue + vuex + webpack + webpack-hot-middleware, It throw cors error.Of course,I have setted webSecurity to false.

**Comment by IAmPicard** [comment]
I uploaded a small repro case here: [https://github.com/IAmPicard/electron-cors-repro](https://github.com/IAmPicard/electron-cors-repro). I tried electron-fiddle, but it looks like the issue doesn't repro when using the "file://" protocol, it needs a local server (http://localhost) such as webpack-dev-server to repro.

**Comment by blackknifes** [comment]
![t](https://user-images.githubusercontent.com/9638801/46987612-862fa800-d127-11e8-8cc6-9476b40ac0e2.png)
![t](https://user-images.githubusercontent.com/9638801/46987633-a52e3a00-d127-11e8-9711-f6f54d559990.png)
![t](https://user-images.githubusercontent.com/9638801/46987700-e888a880-d127-11e8-880b-fb9bf8a09b4e.png)

@IAmPicard 
The project worked as expected on version v2.x, but throw errro on version v3.x.
Why it throw error on version v3.x?

**Comment by IAmPicard** [comment]
@blackknifes I don't believe your issue is related to CORS, it's most likely same-origin, and unrelated to the issue described here (which is CORS-related, and specific to 4.0.0).

**Comment by blackknifes** [comment]
@IAmPicard Ok, I will create new issue.

**Comment by nornagon** [comment]
Confirmed, that repro case behaves differently on 3.0.4 and 4.0.0-beta3.

**Comment by aokihu** [comment]
I have met the same problem, upgrade from 3.0.5 to 4.0-beta, It's good at 3.0.5
My platform is macOS

**Comment by BinaryMuse** [comment]
It looks like `CrossOriginReadBlocking::ResponseAnalyzer::ShouldAllow()` is returning `false` here due to `should_block_based_on_headers_` being set to `kBlock`. 

We were unable in the time we had today to determine if there was a way to affect this behavior based on the `webSecurity` option; however, on `master`, Chromium includes this code in `CrossSiteDocumentResourceHandler::ShouldBlockBasedOnHeaders` (which is what creates and uses the `ResponseAnalyzer`):

```cpp
  // --disable-web-security also disables Cross-Origin Read Blocking (CORB).
  if (base::CommandLine::ForCurrentProcess()->HasSwitch(
          switches::kDisableWebSecurity))
    return false;
```

Unclear to me if this is the best way forward.

**Comment by BinaryMuse** [comment]
After chatting with @deepak1556, https://github.com/electron/electron/pull/15618 should allow us to tackle this in a good way.

**Comment by aokihu** [comment]
Do you fix the problem now? I test beta5, and it does not work well

**Comment by aokihu** [comment]
If the problem can not be fixed, the project is useless from 4.0
An isolate application can not in the same URL domain, so how we communicate with server?

**Comment by deepak1556** [comment]
Although the issue is fixed in https://github.com/electron/electron/pull/15737 , I would suggest not to disable web security in production. Here are some good docs about CORB checks and how to fix it in your application, from the chrome team https://developers.google.com/web/updates/2018/07/site-isolation

For in-depth understanding of CORB https://chromium.googlesource.com/chromium/src/+/master/services/network/cross_origin_read_blocking_explainer.md

**Comment by deepak1556** [comment]
> to move all such network calls down into node 

You can use the [net module](https://github.com/electron/electron/blob/master/docs/api/net.md) along with [stream protocol](https://github.com/electron/electron/blob/master/docs/api/protocol.md#protocolregisterstreamprotocolscheme-handler-completion)

> Or am I missing another alternative that allows us to do fetch() calls from origin file:// against a CORS-enabled server

Yeah I would advice against relying on `file` for these scenarios

**Comment by ckerr** [comment]
Still needs to be backported to `4-0-x`

**Comment by IAmPicard** [comment]
I can confirm the issue is fixed in v4.0.0-beta.8 ; I'm experiencing other unrelated crashes with v4.0.0-beta.8, but this particular regression no longer repros.

Thank you!

**Comment by deepak1556** [comment]
@IAmPicard thanks for confirming.
